File: server/arcade/snake.go
package arcade

import (
	"encoding/json"
	"strconv"

	"cloud-clicker/server/determinism"
	"cloud-clicker/server/minigame"
)

const (
	SnakeEngineRef          = "snake"
	SnakeScalingDestination = "minigame.arcade.snake"
	snakeRunSubstream       = "snake.run.v1"
	snakeFoodSeedSubstream  = "snake.food_seed.v1"
	snakeFoodSubstream      = "snake.food.v1"

	SnakePlaying  = "playing"
	SnakeTerminal = "terminal"

	SnakeCleared = "cleared"
	SnakeCrashed = "crashed"
	SnakeQuit    = "quit"
)

// snakeErrors is AR4.7's sorted rejection taxonomy.
var snakeErrors = []string{"advance_past_terminal", "advance_window", "illegal_phase", "invalid_turn", "turns_not_ascending"}

var snakeFactKinds = []string{"snake.food_eaten", "snake.ticks_survived"}

var snakeDeltas = map[string][2]int64{"up": {0, -1}, "down": {0, 1}, "left": {-1, 0}, "right": {1, 0}}

var snakeOpposites = map[string]string{"up": "down", "down": "up", "left": "right", "right": "left"}

// SnakeSnapshot is AR4.3: exactly fourteen keys. FoodSeed is published as a
// canonical decimal string so an honest client can place food locally.
type SnakeSnapshot struct {
	ArcadeContentHash   string  `json:"arcade_content_hash"`
	ArcadeSchemaVersion int     `json:"arcade_schema_version"`
	Phase               string  `json:"phase"`
	Tick                int64   `json:"tick"`
	Direction           string  `json:"direction"`
	Body                []int64 `json:"body"`
	PendingGrowth       int64   `json:"pending_growth"`
	FoodCell            int64   `json:"food_cell"`
	FoodIndex           int64   `json:"food_index"`
	FoodSeed            string  `json:"food_seed"`
	Score               int64   `json:"score"`
	Width               int64   `json:"width"`
	Height              int64   `json:"height"`
	Revision            int64   `json:"revision"`
}

type SnakeTenant struct{}

func NewSnakeTenant() SnakeTenant { return SnakeTenant{} }

func (SnakeTenant) Descriptor() minigame.Descriptor {
	return minigame.Descriptor{EngineRef: SnakeEngineRef, EngineVersion: EngineVersion,
		CommandSchema: "snake.command.v1", SnapshotSchema: "snake.snapshot.v1", ResultSchema: "minigame.result.v1",
		Modes: []minigame.Mode{minigame.ModeSolo}, ErrorTaxonomy: append([]string(nil), snakeErrors...),
		Destinations: map[string]minigame.DestinationClass{SnakeScalingDestination: minigame.DestinationBreadth}}
}

func (SnakeTenant) ValidateCommand(data json.RawMessage) error {
	_, rejection := decodeSnakeCommand(data)
	return rejection
}

func (SnakeTenant) ValidateSnapshot(data json.RawMessage) error {
	_, err := decodeSnakeSnapshot(data)
	return err
}

func (SnakeTenant) ValidateResult(result *minigame.Result) error {
	if result == nil {
		return nil
	}
	if result.Outcome != SnakeCleared && result.Outcome != SnakeCrashed && result.Outcome != SnakeQuit ||
		result.RatingDelta != nil || len(result.ScoreFacts) != len(snakeFactKinds) {
		return minigame.ErrInvalidTenant
	}
	for index, fact := range result.ScoreFacts {
		if fact.Kind != snakeFactKinds[index] || fact.Value < 0 {
			return minigame.ErrInvalidTenant
		}
	}
	return nil
}

func (SnakeTenant) Create(input minigame.CreateInput) (json.RawMessage, error) {
	catalog, err := catalogForInput(input.Content, input.ContentHash, input.ContentSchemaVersion)
	if err != nil || input.Mode != minigame.ModeSolo || !validLiteralScaling(input.ScalingInputs, SnakeScalingDestination) {
		return nil, minigame.ErrInvalidTenant
	}
	snapshot := SnakeGenesis(catalog.Snake, input.Seed)
	snapshot.ArcadeContentHash, snapshot.ArcadeSchemaVersion = input.ContentHash, input.ContentSchemaVersion
	return encodeCanonical(snapshot)
}

// SnakeGenesis is AR4.3: head at (width/2, height/2), the body extending
// west to start_length, direction right, and food 0 placed.
func SnakeGenesis(content SnakeContent, seed uint64) SnakeSnapshot {
	head := (content.Height/2)*content.Width + content.Width/2
	body := make([]int64, 0, content.StartLength)
	for offset := int64(0); offset < content.StartLength; offset++ {
		body = append(body, head-offset)
	}
	foodSeed := SnakeFoodSeed(seed)
	snapshot := SnakeSnapshot{Phase: SnakePlaying, Direction: "right", Body: body, FoodSeed: strconv.FormatUint(foodSeed, 10),
		Width: content.Width, Height: content.Height, Revision: 1}
	snapshot.FoodCell = placeFood(snapshot.Body, content.Width*content.Height, foodSeed, 0)
	snapshot.FoodIndex = 1
	return snapshot
}

// SnakeFoodSeed is AR4.2's published food seed for a session seed.
func SnakeFoodSeed(seed uint64) uint64 {
	runSeed := determinism.Substream(seed, snakeRunSubstream).Next()
	return determinism.Substream(runSeed, snakeFoodSeedSubstream).Next()
}

// placeFood puts food number k on the ascending list of empty cells, or
// returns -1 when no empty cell remains.
func placeFood(body []int64, cells int64, foodSeed uint64, k int64) int64 {
	occupied := int64Set(body)
	empty := make([]int64, 0, cells)
	for cell := int64(0); cell < cells; cell++ {
		if !occupied[cell] {
			empty = append(empty, cell)
		}
	}
	if len(empty) == 0 {
		return -1
	}
	return empty[determinism.Substream(foodSeed^uint64(k), snakeFoodSubstream).Bound(uint64(len(empty)))]
}

func (SnakeTenant) Apply(input minigame.ApplyInput) (minigame.ApplyOutput, error) {
	catalog, err := catalogForInput(input.Content, input.ContentHash, input.ContentSchemaVersion)
	if err != nil || input.Mode != minigame.ModeSolo || !validLiteralScaling(input.ScalingInputs, SnakeScalingDestination) {
		return minigame.ApplyOutput{}, minigame.ErrInvalidTenant
	}
	snapshot, err := decodeSnakeSnapshot(input.Snapshot)
	if err != nil || snapshot.Revision != input.Revision || snapshot.ArcadeContentHash != input.ContentHash ||
		snapshot.ArcadeSchemaVersion != input.ContentSchemaVersion || snapshot.Width != catalog.Snake.Width ||
		snapshot.Height != catalog.Snake.Height || snapshot.FoodSeed != strconv.FormatUint(SnakeFoodSeed(input.Seed), 10) {
		return minigame.ApplyOutput{}, minigame.ErrTenantDivergence
	}
	command, rejection := decodeSnakeCommand(input.Command)
	if rejection != nil {
		return minigame.ApplyOutput{}, rejection
	}
	result, rejection := snakeTransition(&snapshot, command, catalog.Snake)
	if rejection != nil {
		return minigame.ApplyOutput{}, rejection
	}
	snapshot.Revision = input.Revision + 1
	encoded, err := encodeCanonical(snapshot)
	if err != nil {
		return minigame.ApplyOutput{}, minigame.ErrTenantDivergence
	}
	return minigame.ApplyOutput{Snapshot: encoded, Result: result}, nil
}

type snakeTurn struct {
	Tick      int64
	Direction string
}

type snakeCommand struct {
	Kind        string
	ThroughTick int64
	Turns       []snakeTurn
}

// snakeTransition simulates on a copy so a rejected command mutates nothing.
func snakeTransition(snapshot *SnakeSnapshot, command snakeCommand, content SnakeContent) (*minigame.Result, error) {
	if snapshot.Phase != SnakePlaying {
		return nil, reject("illegal_phase", command.Kind+" requires playing phase")
	}
	if command.Kind == "quit" {
		return terminateSnake(snapshot, SnakeQuit), nil
	}
	if command.ThroughTick <= snapshot.Tick || command.ThroughTick > snapshot.Tick+content.MaxTicksPerAdvance {
		return nil, reject("advance_window", "through_tick is outside the advance window")
	}
	for index, turn := range command.Turns {
		if turn.Tick <= snapshot.Tick || turn.Tick > command.ThroughTick {
			return nil, reject("advance_window", "turn tick is outside the advance window")
		}
		if index > 0 && command.Turns[index-1].Tick >= turn.Tick {
			return nil, reject("turns_not_ascending", "turn ticks must be strictly ascending")
		}
	}
	work := cloneSnake(*snapshot)
	foodSeed, _ := strconv.ParseUint(work.FoodSeed, 10, 64)
	cells := work.Width * work.Height
	turnIndex := 0
	for t := work.Tick + 1; t <= command.ThroughTick; t++ {
		if turnIndex < len(command.Turns) && command.Turns[turnIndex].Tick == t {
			next := command.Turns[turnIndex].Direction
			if next == work.Direction || next == snakeOpposites[work.Direction] {
				return nil, reject("invalid_turn", "a turn must be a 90 degree change")
			}
			work.Direction = next
			turnIndex++
		}
		outcome := snakeStep(&work, t, cells, content.GrowthPerFood, foodSeed)
		if outcome != "" {
			if t != command.ThroughTick {
				return nil, reject("advance_past_terminal", "the game ended before through_tick")
			}
			*snapshot = work
			return terminateSnake(snapshot, outcome), nil
		}
	}
	*snapshot = work
	return nil, nil
}

// snakeStep is AR4.4 in its exact order; it returns a terminal outcome or "".
func snakeStep(snapshot *SnakeSnapshot, t, cells, growthPerFood int64, foodSeed uint64) string {
	delta := snakeDeltas[snapshot.Direction]
	head := snapshot.Body[0]
	x, y := head%snapshot.Width+delta[0], head/snapshot.Width+delta[1]
	if x < 0 || y < 0 || x >= snapshot.Width || y >= snapshot.Height {
		snapshot.Tick = t
		return SnakeCrashed
	}
	next := y*snapshot.Width + x
	growing := snapshot.PendingGrowth > 0
	kept := snapshot.Body
	if !growing {
		kept = snapshot.Body[:len(snapshot.Body)-1]
	}
	for _, cell := range kept {
		if cell == next {
			snapshot.Tick = t
			return SnakeCrashed
		}
	}
	snapshot.Body = append([]int64{next}, kept...)
	if growing {
		snapshot.PendingGrowth--
	}
	if next == snapshot.FoodCell {
		snapshot.Score++
		snapshot.PendingGrowth += growthPerFood
		snapshot.FoodCell = placeFood(snapshot.Body, cells, foodSeed, snapshot.FoodIndex)
		snapshot.FoodIndex++
		if snapshot.FoodCell == -1 {
			snapshot.Tick = t
			return SnakeCleared
		}
	}
	snapshot.Tick = t
	return ""
}

func terminateSnake(snapshot *SnakeSnapshot, outcome string) *minigame.Result {
	snapshot.Phase = SnakeTerminal
	return &minigame.Result{Outcome: outcome, RatingDelta: nil, ScoreFacts: []minigame.ScoreFact{
		{Kind: "snake.food_eaten", Value: snapshot.Score},
		{Kind: "snake.ticks_survived", Value: snapshot.Tick},
	}}
}

func cloneSnake(value SnakeSnapshot) SnakeSnapshot {
	value.Body = append([]int64(nil), value.Body...)
	return value
}

func decodeSnakeCommand(data []byte) (snakeCommand, error) {
	if !uniqueJSONKeys(data) {
		return snakeCommand{}, reject("illegal_phase", "command keys are not unique")
	}
	var header struct {
		Kind string `json:"kind"`
	}
	if json.Unmarshal(data, &header) != nil {
		return snakeCommand{}, reject("illegal_phase", "command is not an object")
	}
	switch header.Kind {
	case "advance":
		var wire struct {
			Kind        string            `json:"kind"`
			ThroughTick json.RawMessage   `json:"through_tick"`
			Turns       []json.RawMessage `json:"turns"`
		}
		if !hasExactJSONKeys(data, "kind", "through_tick", "turns") || strictDecode(data, &wire) != nil || wire.Turns == nil {
			return snakeCommand{}, reject("advance_window", "advance schema mismatch")
		}
		through, ok := exactInteger(wire.ThroughTick)
		if !ok {
			return snakeCommand{}, reject("advance_window", "through_tick must be a safe integer")
		}
		command := snakeCommand{Kind: header.Kind, ThroughTick: through, Turns: make([]snakeTurn, 0, len(wire.Turns))}
		for _, raw := range wire.Turns {
			var turn struct {
				Tick      json.RawMessage `json:"tick"`
				Direction string          `json:"direction"`
			}
			if !hasExactJSONKeys(raw, "tick", "direction") || strictDecode(raw, &turn) != nil {
				return snakeCommand{}, reject("advance_window", "turn schema mismatch")
			}
			tick, tickOK := exactInteger(turn.Tick)
			if !tickOK {
				return snakeCommand{}, reject("advance_window", "turn tick must be a safe integer")
			}
			if _, known := snakeDeltas[turn.Direction]; !known {
				return snakeCommand{}, reject("invalid_turn", "unknown direction")
			}
			command.Turns = append(command.Turns, snakeTurn{Tick: tick, Direction: turn.Direction})
		}
		return command, nil
	case "quit":
		if !hasExactJSONKeys(data, "kind") {
			return snakeCommand{}, reject("illegal_phase", "quit schema mismatch")
		}
		return snakeCommand{Kind: header.Kind}, nil
	default:
		return snakeCommand{}, reject("illegal_phase", "unknown command kind")
	}
}

func decodeSnakeSnapshot(data []byte) (SnakeSnapshot, error) {
	if !uniqueJSONKeys(data) || !hasExactJSONKeys(data, "arcade_content_hash", "arcade_schema_version", "phase", "tick", "direction", "body",
		"pending_growth", "food_cell", "food_index", "food_seed", "score", "width", "height", "revision") {
		return SnakeSnapshot{}, minigame.ErrInvalidTenant
	}
	var value SnakeSnapshot
	if strictDecode(data, &value) != nil || value.ArcadeSchemaVersion != SchemaVersion || !validHash(value.ArcadeContentHash) ||
		value.Phase != SnakePlaying && value.Phase != SnakeTerminal || value.Revision < 1 || value.Tick < 0 ||
		value.Width < 5 || value.Width > 30 || value.Height < 5 || value.Height > 30 || len(value.Body) < 2 ||
		value.PendingGrowth < 0 || value.Score < 0 || value.FoodIndex < 1 || value.FoodIndex != value.Score+1 {
		return SnakeSnapshot{}, minigame.ErrInvalidTenant
	}
	if _, ok := snakeDeltas[value.Direction]; !ok {
		return SnakeSnapshot{}, minigame.ErrInvalidTenant
	}
	parsed, err := strconv.ParseUint(value.FoodSeed, 10, 64)
	if err != nil || strconv.FormatUint(parsed, 10) != value.FoodSeed {
		return SnakeSnapshot{}, minigame.ErrInvalidTenant
	}
	cells := value.Width * value.Height
	seen := map[int64]bool{}
	for _, cell := range value.Body {
		if cell < 0 || cell >= cells || seen[cell] {
			return SnakeSnapshot{}, minigame.ErrInvalidTenant
		}
		seen[cell] = true
	}
	if value.FoodCell < -1 || value.FoodCell >= cells || value.FoodCell >= 0 && seen[value.FoodCell] ||
		value.FoodCell == -1 && value.Phase != SnakeTerminal {
		return SnakeSnapshot{}, minigame.ErrInvalidTenant
	}
	return value, nil
}

File: client/src/arcade/snake.ts
import { substream } from "../combat/rng";
import type { ArcadeSnakeContent } from "./catalog";
import { ArcadeRejection, encodeCanonical, keysOf, parseCommandObject, requireLiteralScaling, resolveArcadeCatalog, safeInteger,
  type ArcadeApplyInput, type ArcadeCreateInput, type ArcadeResult } from "./common";

// `snake` 1.0.0 (AR4), the TS mirror of server/arcade/snake.go. The client
// also runs `snakeStep` locally at presentation pace (AR6.3); the server
// snapshot stays the only truth.
export const SNAKE_ENGINE_REF = "snake" as const;
export const SNAKE_SCALING_DESTINATION = "minigame.arcade.snake" as const;
const RUN_SUBSTREAM = "snake.run.v1";
const FOOD_SEED_SUBSTREAM = "snake.food_seed.v1";
const FOOD_SUBSTREAM = "snake.food.v1";

export type SnakeDirection = "up" | "down" | "left" | "right";
export type SnakePhase = "playing" | "terminal";
export interface SnakeSnapshot {
  readonly arcade_content_hash: string; readonly arcade_schema_version: number; readonly phase: SnakePhase; readonly tick: number;
  readonly direction: SnakeDirection; readonly body: readonly number[]; readonly pending_growth: number; readonly food_cell: number;
  readonly food_index: number; readonly food_seed: string; readonly score: number; readonly width: number; readonly height: number; readonly revision: number;
}
export interface SnakeTurn { readonly tick: number; readonly direction: SnakeDirection }
export type SnakeCommand = Readonly<{ kind: "advance"; through_tick: number; turns: readonly SnakeTurn[] }> | Readonly<{ kind: "quit" }>;
export type SnakeMutable = { -readonly [K in keyof SnakeSnapshot]: SnakeSnapshot[K] };

const DELTAS: Readonly<Record<SnakeDirection, readonly [number, number]>> = { up: [0, -1], down: [0, 1], left: [-1, 0], right: [1, 0] };
const OPPOSITES: Readonly<Record<SnakeDirection, SnakeDirection>> = { up: "down", down: "up", left: "right", right: "left" };

export function isSnakeDirection(value: unknown): value is SnakeDirection { return value === "up" || value === "down" || value === "left" || value === "right"; }
export function snakeOpposite(direction: SnakeDirection): SnakeDirection { return OPPOSITES[direction]; }

export function snakeFoodSeed(seed: bigint): bigint {
  const runSeed = substream(seed, RUN_SUBSTREAM).next();
  return substream(runSeed, FOOD_SEED_SUBSTREAM).next();
}

/** Food k on the ascending list of empty cells, or -1 when the board is full. */
export function snakePlaceFood(body: readonly number[], cells: number, foodSeed: bigint, k: number): number {
  const occupied = new Set(body), empty: number[] = [];
  for (let cell = 0; cell < cells; cell++) if (!occupied.has(cell)) empty.push(cell);
  if (empty.length === 0) return -1;
  return empty[Number(substream(foodSeed ^ BigInt(k), FOOD_SUBSTREAM).bound(BigInt(empty.length)))]!;
}

export async function createSnake(input: ArcadeCreateInput): Promise<string> {
  const catalog = await resolveArcadeCatalog(input);
  if (input.mode !== "solo") throw new SyntaxError("snake is solo only");
  requireLiteralScaling(input.scaling_inputs, SNAKE_SCALING_DESTINATION);
  return encodeCanonical({ ...snakeGenesis(catalog.snake, input.seed), arcade_content_hash: input.content_hash, arcade_schema_version: input.content_schema_version });
}

export function snakeGenesis(content: ArcadeSnakeContent, seed: bigint): Omit<SnakeSnapshot, "arcade_content_hash" | "arcade_schema_version"> {
  const head = Math.floor(content.height / 2) * content.width + Math.floor(content.width / 2);
  const body = Array.from({ length: content.start_length }, (_, offset) => head - offset);
  const foodSeed = snakeFoodSeed(seed);
  return { phase: "playing", tick: 0, direction: "right", body, pending_growth: 0, food_cell: snakePlaceFood(body, content.width * content.height, foodSeed, 0),
    food_index: 1, food_seed: foodSeed.toString(), score: 0, width: content.width, height: content.height, revision: 1 };
}

export async function applySnake(input: ArcadeApplyInput): Promise<{ readonly snapshot: string; readonly result: ArcadeResult | null }> {
  const catalog = await resolveArcadeCatalog(input);
  requireLiteralScaling(input.scaling_inputs, SNAKE_SCALING_DESTINATION);
  const snapshot = decodeSnakeSnapshot(input.snapshot);
  if (snapshot.revision !== input.revision || snapshot.arcade_content_hash !== input.content_hash || snapshot.arcade_schema_version !== input.content_schema_version ||
    snapshot.width !== catalog.snake.width || snapshot.height !== catalog.snake.height || snapshot.food_seed !== snakeFoodSeed(input.seed).toString()) {
    throw new SyntaxError("snake snapshot diverges from its identity");
  }
  const command = decodeSnakeCommand(input.command);
  const { snapshot: next, result } = snakeTransition(snapshot, command, catalog.snake);
  next.revision = input.revision + 1;
  return { snapshot: encodeCanonical(next), result };
}

/** Pure transition on a copy: a rejected command mutates nothing (AR4.5). */
export function snakeTransition(snapshot: SnakeSnapshot, command: SnakeCommand, content: ArcadeSnakeContent): { snapshot: SnakeMutable; result: ArcadeResult | null } {
  if (snapshot.phase !== "playing") throw new ArcadeRejection("illegal_phase", `${command.kind} requires playing phase`);
  const work: SnakeMutable = { ...snapshot, body: [...snapshot.body] };
  if (command.kind === "quit") return { snapshot: work, result: terminate(work, "quit") };
  if (command.through_tick <= snapshot.tick || command.through_tick > snapshot.tick + content.max_ticks_per_advance) throw new ArcadeRejection("advance_window", "through_tick is outside the advance window");
  command.turns.forEach((turn, index) => {
    if (turn.tick <= snapshot.tick || turn.tick > command.through_tick) throw new ArcadeRejection("advance_window", "turn tick is outside the advance window");
    if (index > 0 && command.turns[index - 1]!.tick >= turn.tick) throw new ArcadeRejection("turns_not_ascending", "turn ticks must be strictly ascending");
  });
  const foodSeed = BigInt(work.food_seed);
  let turnIndex = 0;
  for (let t = work.tick + 1; t <= command.through_tick; t++) {
    const turn = command.turns[turnIndex];
    if (turn && turn.tick === t) {
      if (turn.direction === work.direction || turn.direction === OPPOSITES[work.direction]) throw new ArcadeRejection("invalid_turn", "a turn must be a 90 degree change");
      work.direction = turn.direction;
      turnIndex++;
    }
    const outcome = snakeStep(work, t, content.growth_per_food, foodSeed);
    if (outcome !== null) {
      if (t !== command.through_tick) throw new ArcadeRejection("advance_past_terminal", "the game ended before through_tick");
      return { snapshot: work, result: terminate(work, outcome) };
    }
  }
  return { snapshot: work, result: null };
}

/** AR4.4 in its exact order; returns a terminal outcome or null. */
export function snakeStep(snapshot: SnakeMutable, t: number, growthPerFood: number, foodSeed: bigint): "cleared" | "crashed" | null {
  const [dx, dy] = DELTAS[snapshot.direction];
  const head = snapshot.body[0]!;
  const x = head % snapshot.width + dx, y = Math.floor(head / snapshot.width) + dy;
  if (x < 0 || y < 0 || x >= snapshot.width || y >= snapshot.height) { snapshot.tick = t; return "crashed"; }
  const next = y * snapshot.width + x;
  const growing = snapshot.pending_growth > 0;
  const kept = growing ? snapshot.body : snapshot.body.slice(0, -1);
  if (kept.includes(next)) { snapshot.tick = t; return "crashed"; }
  snapshot.body = [next, ...kept];
  if (growing) snapshot.pending_growth--;
  if (next === snapshot.food_cell) {
    snapshot.score++;
    snapshot.pending_growth += growthPerFood;
    snapshot.food_cell = snakePlaceFood(snapshot.body, snapshot.width * snapshot.height, foodSeed, snapshot.food_index);
    snapshot.food_index++;
    if (snapshot.food_cell === -1) { snapshot.tick = t; return "cleared"; }
  }
  snapshot.tick = t;
  return null;
}

function terminate(snapshot: SnakeMutable, outcome: "cleared" | "crashed" | "quit"): ArcadeResult {
  snapshot.phase = "terminal";
  return { outcome, rating_delta: null, score_facts: [{ kind: "snake.food_eaten", value: snapshot.score }, { kind: "snake.ticks_survived", value: snapshot.tick }] };
}

export function decodeSnakeCommand(source: string): SnakeCommand {
  const row = parseCommandObject(source, () => { throw new ArcadeRejection("illegal_phase", "command is not an object"); });
  if (row.kind === "advance") {
    if (keysOf(row) !== "kind\0through_tick\0turns" || !Array.isArray(row.turns)) throw new ArcadeRejection("advance_window", "advance schema mismatch");
    if (!safeInteger(row.through_tick)) throw new ArcadeRejection("advance_window", "through_tick must be a safe integer");
    const turns = row.turns.map((item): SnakeTurn => {
      if (item === null || typeof item !== "object" || Array.isArray(item) || keysOf(item as Record<string, unknown>) !== "direction\0tick") throw new ArcadeRejection("advance_window", "turn schema mismatch");
      const turn = item as Record<string, unknown>;
      if (!safeInteger(turn.tick) || typeof turn.direction !== "string") throw new ArcadeRejection("advance_window", "turn tick must be a safe integer and direction a string");
      if (!isSnakeDirection(turn.direction)) throw new ArcadeRejection("invalid_turn", "unknown direction");
      return { tick: turn.tick, direction: turn.direction };
    });
    return { kind: "advance", through_tick: row.through_tick, turns };
  }
  if (row.kind === "quit") {
    if (keysOf(row) !== "kind") throw new ArcadeRejection("illegal_phase", "quit schema mismatch");
    return { kind: "quit" };
  }
  throw new ArcadeRejection("illegal_phase", "unknown command kind");
}

export function decodeSnakeSnapshot(source: string): SnakeMutable {
  const value = JSON.parse(source) as SnakeMutable;
  const keys = ["arcade_content_hash", "arcade_schema_version", "body", "direction", "food_cell", "food_index", "food_seed", "height", "pending_growth", "phase",
    "revision", "score", "tick", "width"];
  if (value === null || typeof value !== "object" || Object.keys(value).sort().join("\0") !== keys.join("\0") || value.arcade_schema_version !== 1 ||
    !/^sha256:[0-9a-f]{64}$/.test(value.arcade_content_hash) || value.phase !== "playing" && value.phase !== "terminal" || value.revision < 1 || value.tick < 0 ||
    value.width < 5 || value.width > 30 || value.height < 5 || value.height > 30 || !Array.isArray(value.body) || value.body.length < 2 ||
    value.pending_growth < 0 || value.score < 0 || value.food_index < 1 || value.food_index !== value.score + 1 || !isSnakeDirection(value.direction) ||
    !/^(0|[1-9][0-9]*)$/.test(value.food_seed) || BigInt(value.food_seed) >= 1n << 64n) throw new SyntaxError("invalid snake snapshot");
  const cells = value.width * value.height, seen = new Set<number>();
  for (const cell of value.body) {
    if (!Number.isSafeInteger(cell) || cell < 0 || cell >= cells || seen.has(cell)) throw new SyntaxError("invalid snake body");
    seen.add(cell);
  }
  if (value.food_cell < -1 || value.food_cell >= cells || value.food_cell >= 0 && seen.has(value.food_cell) || value.food_cell === -1 && value.phase !== "terminal") {
    throw new SyntaxError("invalid snake food");
  }
  return value;
}
